tests/unit/test_dso_transport.cpp

```diff
@@ -0,0 +1,364 @@
+    {
+        int length = aData.size();
+
+        aData.insert(aData.begin(), length >> 8);
+        aData.insert(aData.begin() + 1, length & ((1 << 8) - 1));
+
+        return aData;
+    }
+
+    static void Send(mbedtls_net_context &aCtx, const std::vector<uint8_t> &aData)
+    {
+        auto data = MakeDataForSending(aData);
+
+        mbedtls_net_send(&aCtx, data.data(), data.size());
+    }
+
+    static std::vector<uint8_t> Receive(mbedtls_net_context &aCtx)
+    {
+        uint16_t             length;
+        std::vector<uint8_t> data;
+
+        CHECK(sizeof(length) == mbedtls_net_recv(&aCtx, reinterpret_cast<uint8_t *>(&length), sizeof(length)));
+        length = ntohs(length);
+        data.resize(length);
+        CHECK(length == mbedtls_net_recv(&aCtx, data.data(), length));
+
+        return data;
+    }
+
+    void SetUp()
+    {
+        mAgent.Init(mInstance, "lo");
+        mAgent.SetHandlers(
+            [this](otInstance *aInstance, otSockAddr *aSockAddr) { return AcceptConnection(aInstance, aSockAddr); },
+            [this](otbr::Dso::DsoAgent::DsoConnection *aConnection) { HandleConnected(aConnection); },
+            [this](otbr::Dso::DsoAgent::DsoConnection *aConnection, otPlatDsoDisconnectMode aMode) {
+                HandleDisconnected(aConnection, aMode);
+            },
+            [this](otbr::Dso::DsoAgent::DsoConnection *aConnection, const std::vector<uint8_t> &aData) {
+                HandleReceive(aConnection, aData);
+            });
+    }
+
+    std::function<void(void)> RunMainLoop()
+    {
+        mShouldExit                 = false;
+        std::thread *mainLoopThread = new std::thread([this] {
+            while (!mShouldExit)
+            {
+                int                   rval;
+                otbr::MainloopContext mainloop{};
+
+                mainloop.mMaxFd           = -1;
+                mainloop.mTimeout.tv_sec  = 1; // 1 second
+                mainloop.mTimeout.tv_usec = 0;
+
+                FD_ZERO(&mainloop.mReadFdSet);
+                FD_ZERO(&mainloop.mWriteFdSet);
+                FD_ZERO(&mainloop.mErrorFdSet);
+
+                otbr::MainloopManager::GetInstance().Update(mainloop);
+
+                rval = select(mainloop.mMaxFd + 1, &mainloop.mReadFdSet, &mainloop.mWriteFdSet, &mainloop.mErrorFdSet,
+                              &mainloop.mTimeout);
+
+                if (rval >= 0)
+                {
+                    otbr::MainloopManager::GetInstance().Process(mainloop);
+                }
+            }
+        });
+
+        return [this, mainLoopThread]() {
+            mShouldExit = true;
+            mainLoopThread->join();
+            delete mainLoopThread;
+        };
+    }
+
+    void RunMainLoopFor(const std::chrono::milliseconds &aMilliseconds)
+    {
+        auto quitMainLoop = RunMainLoop();
+
+        std::this_thread::sleep_for(aMilliseconds);
+
+        quitMainLoop();
+    }
+
+    void TestServer(void)
```

Comment:
Try adding failure test cases where client/server dies in the middle to see if the Agent can handle those error cases and recycle connections properly.